smartapp-spring/src/main/java/com/smartthings/sdk/smartapp/spring/HttpVerificationService.java

```diff
@@ -0,0 +1,140 @@
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.servlet.http.HttpServletRequest;
+
+import org.bouncycastle.asn1.x509.SubjectPublicKeyInfo;
+import org.bouncycastle.crypto.params.RSAKeyParameters;
+import org.bouncycastle.crypto.util.PublicKeyFactory;
+import org.bouncycastle.openssl.PEMParser;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.core.io.Resource;
+import org.springframework.util.FileCopyUtils;
+
+import net.adamcin.httpsig.api.Authorization;
+import net.adamcin.httpsig.api.Challenge;
+import net.adamcin.httpsig.api.DefaultKeychain;
+import net.adamcin.httpsig.api.DefaultVerifier;
+import net.adamcin.httpsig.api.KeyId;
+import net.adamcin.httpsig.api.RequestContent;
+import net.adamcin.httpsig.api.Verifier;
+import net.adamcin.httpsig.ssh.jce.KeyFormat;
+import net.adamcin.httpsig.ssh.jce.SSHKey;
+import net.adamcin.httpsig.ssh.jce.UserFingerprintKeyId;
+
+
+/**
+ * All requests should have their HTTP signature verified to ensure if request is actually from SmartThings.
+ * This class has the implementation of signature authentication and verification
+ * from https://github.com/adamcin/httpsig-java.
+ * The public key that is generated when registering a smartapp; is used to verify the request.
+ */
+public class HttpVerificationService {
+    private static final Logger LOG = LoggerFactory.getLogger(HttpVerificationService.class);
+
+    private final String publicKey;
+
+    public HttpVerificationService(Resource publicKeyResource) {
+        if (!publicKeyResource.exists()) {
+            // This is not an error until we are past the PING step.
+            LOG.info("No public key yet; will only accept PING lifecycle requests." +
+                    " Create Automation in Developer workspace to continue.");
+            publicKey = null;
+            return;
+        }
+
+        try {
+            LOG.debug("Using public key file: " + publicKeyResource.getFilename());
+            publicKey = new String(FileCopyUtils.copyToByteArray(publicKeyResource.getInputStream()), "UTF-8");
+        } catch (IOException ioException) {
+            LOG.error("Could not read public key file", ioException);
+            throw new RuntimeException("Could not read public key file", ioException);
+        }
+    }
+
+    public boolean verify(HttpServletRequest request) {
+        if (publicKey == null) {
+            LOG.error("Public key file not set up properly. Please see README (Configure Public Key) for"
+                    + " directions and don't forget to restart this server.");
+            return false;
+        }
+        boolean verified = verifyRequest(request);
+        if (!verified) {
+            LOG.error("Request not verified!");
+        }
+        return verified;
+    }
+
+    private boolean verifyRequest(HttpServletRequest request) {
```

Comment:
Are there any unchecked exceptions to be mindful of here?

Reply:
I think the error handling here is good.